### PR TITLE
add(microbial fastq order submitter)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Backend solution part of : https://github.com/Clinical-Genomics/cg/issues/3338

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: tests/fixtures/cgweb_orders/microbial_fastq.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            "data_delivery": "fastq",
+            "elution_buffer": "Nuclease-free water",
+            "name": "prov2",
+            "priority": "priority",
+            "sex": "male",
+            "volume": "2",
+            "well_position": ""
+        }
+    ]
+}
```
Comment (ChrOertlin) [inline tests/fixtures/cgweb_orders/microbial_fastq.json:35]: @Vince-janv I took a fastq order and modified it a bit, could you have a look whether this is correct here? Then I will add tests around this.

File: tests/fixtures/cgweb_orders/microbial_fastq.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            "application": "WGSPCFC060",
+            "comment": "sample comment",
+            "container": "Tube",
+            "container_name": "prov2",
+            "data_analysis": "fastq",
+            "data_delivery": "fastq",
+            "elution_buffer": "Nuclease-free water",
+            "name": "prov2",
+            "priority": "priority",
+            "sex": "male",
```
Comment (Vince-janv) [inline tests/fixtures/cgweb_orders/microbial_fastq.json:30]: Sex is not relevant here

File: cg/utils/error_handler.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            try:
+                return fn(*args, **kwargs)
+            except to_except as error:
+                raise to_raise(error) from error
+            except Exception as error:
+                raise CgError(f"{error}") from error
+
+        return wrapper
+
+    return error_handling
```
Comment (Vince-janv) [inline cg/utils/error_handler.py:20]: We had a discussion during your vacation about these error handlers and we all liked them, but though the general ones where you pass exceptions to it where were not as general as they seem. I would prefer having this function in the orders module and specifying in the method itself what to catch and raise

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        customer_id: str,
+        order: str,
+        ordered: datetime,
+        ticket_id: str,
+        items: list[dict],
+    ) -> list[Sample]:
+        customer: Customer = self._get_customer(customer_id)
+        new_samples: list[Sample] = []
+        for sample in items:
+            case_name: str = sample["name"]
```
Comment (Vince-janv) [inline cg/meta/orders/microbial_fastq_submitter.py:64]: Will this get confusing? Should we add a prefix or something to have it different than the sample name?

**Review by Vince-janv — APPROVED** [review]
Great work! ⚡

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    "data_analysis": sample.data_analysis,
+                    "data_delivery": sample.data_delivery,
+                    "name": sample.name,
+                    "priority": sample.priority,
+                    "volume": sample.volume,
+                    "control": sample.control,
+                }
+                for sample in order.samples
+            ],
+        }
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:51]: Why do we need this dict? Is it not easier to work with the BaseModel?

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        status_data: dict = self.order_to_status(order)
+        self._fill_in_sample_ids(samples=status_data["samples"], lims_map=lims_map)
+        new_samples: list[Sample] = self.store_items_in_status(
+            customer_id=status_data["customer"],
+            order=status_data["order"],
+            ordered=project_data["date"] if project_data else datetime.now(),
+            ticket_id=order.ticket,
+            items=status_data["samples"],
+        )
+        self._add_missing_reads(new_samples)
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:29]: I guess the `_add_missing_reads` should be removed? Or did we not have a PR open for removing it?

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        ordered: datetime,
+        ticket_id: str,
+        items: list[dict],
+    ) -> list[Sample]:
+        customer: Customer = self._get_customer(customer_id)
+        new_samples: list[Sample] = []
+        for sample in items:
+            case_name: str = sample["name"]
+            case: Case = self._create_case_for_sample(sample, customer, case_name)
+            db_sample: Sample = self._create_sample(sample, order, ordered, ticket_id)
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:66]: Named arguments

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        raise EntryNotFoundError(f"could not find customer: {customer_id}")
+
+    def _create_case_for_sample(self, sample: dict, customer: Customer, case_name: str) -> Case:
+        if case := self.status.get_case_by_name_and_customer(
+            case_name=case_name, customer=customer
+        ):
+            return case
+        case: Case = self.status.add_case(
+            data_analysis=Workflow.FASTQ,
+            data_delivery=DataDelivery.FASTQ,
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:90]: Data delivery should be picked from the order. There is also a "no delivery" option.

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            name=case_name,
+            panels=None,
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:92]:
```suggestion
```

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            priority=sample["priority"],
+            ticket=case_name,
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:94]: This seems weird to me - should the ticket not be the SuSy ticket_id?

File: cg/meta/orders/microbial_fastq_submitter.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+from cg.meta.orders.lims import process_lims
+from cg.meta.orders.submitter import Submitter
+from cg.models.orders.order import OrderIn
+from cg.models.orders.sample_base import StatusEnum
+from cg.store.exc import EntryNotFoundError
+from cg.store.models import Sample, Customer, Case, CaseSample
+from cg.utils.error_handler import handle_errors
+
+
+class MicrobialFastqSubmitter(Submitter):
```
Comment (islean) [inline cg/meta/orders/microbial_fastq_submitter.py:14]: I'm wondering if we are missing logic for existing samples? Usually these are identified by already having an internal_id filled in. Such samples should not be added to StatusDB and to my knowledge we would only need to connect them to the generated case.

**Review by islean — COMMENTED** [review]
I think we are missing some important logic for existing cases, and I do not like the dict being passed around. But apart from that it looks good!

**Comment by ChrOertlin** [comment]
Closing this as submitters have been refactored - will lnk this PR to salvage functions for new pr

**Pull request closed without merging**